### ISSUE TITLE
Sizeof doesn't work with PascalString when given a context

### ISSUE THREAD

**Issue by Schamper** [construct/construct#268]
Looks like it's because it's using `GreedyBytes` underwater. However, I feel like given a context PascalString should be able to return its length (that being the length of the string). I suppose the same could be said for CString.

Let me know if you want a unit test added for this.

**Comment by arekbulski** [comment]
No, sizeof returns an int on fixed size constructs and SizeofError on variable length. That means `CString PascalString GreedyString` are all variable length. `String` is different, its fixed length and it uses padding and timming. If you want to measure size for a particular string, you need to use
`len(PascalString(VarInt).build(?))`

**Comment by Schamper** [comment]
But what about context though? I know a variable length field can't have a fixed size, but I think it'd be tremendously useful if variable length fields could return a size when given a context to work with.

In my current situation, I have a struct split up into 3 nested structs: `header`, `payload0` and `payload1`. In the header I need some size fields for the size of both payloads. Ideally I'd just be able to do `payload0.sizeof(context)`.

**Comment by arekbulski** [comment]
I think you should look at `RawCopy` wrapper.

**Comment by Schamper** [comment]
I'm not too keen on `RawCopy` since it forces you to traverse an additional `value` key.

Rebuild is what I intended to use, but I can't use it because I can't do `this.payload0.sizeof(context)` ;) (because `payload0` contains PascalString's)
Doing something like `len(payload0.build(context))` seems like a waste of CPU cycles since it'll be built right after _again_, but I suppose it's my only option of solving this exclusively with contruct at the moment.

**Comment by arekbulski** [comment]
You could just build the payloads in separate structures, then use outter struct that just contains `Prefixed` or `GreedyBytes`.

**Comment by Schamper** [comment]
Yeah that was my original plan, but I wanted to try and see if I could solve all my parsing/building needs exclusively with construct.

**Comment by Schamper** [comment]
Been trying to work around this issue but am not quite happy with the result so far. My goal for using Construct was to get rid of the rather complex packing logic I had and replace it with Construct's Switches and the like. To achieve what I want I'd need to deal with a whole lot of hackery, which I'd like to keep to a minimum.

In my opinion, giving a subcontruct access to the `obj` in `_sizeof` (the same `obj` that would be passed into `_build`), allowing a subconstruct to calculate the relevant size to that particular context, would solve this issue. The docstring for `sizeof` even says:
```
Some constructs have no fixed size and can only know their size for a given hunk of data. These constructs will raise an error if they are not passed a context.
```
Reading that, I'd expected that if I'd pass the exact same `Container` into `sizeof` as I'd do into `build`, it'll give me the same size as when I'd do `len(blah.build(x))`. 

I realize that this isn't an easy change. You already provided alternatives to my particular issue (which I'd like to thank you for, but I'd rather try and tackle this issue in the most elegant, contruct-y way possible), but I'd like your opinion and arguments on my point of view.

Slightly related to #266, since my "method" also has to do with traversing contexts inside of `sizeof`.

**Comment by arekbulski** [comment]
I think the `RawCopy` is the most straightforward way to go about anything that cannot be solved more directly with `Prefixed`. Yeah, you get the useless `dict(value=...)` but that isnt much inconvenience compared to other solutions.

When I took over the project I thought exactly the same thing, sizeof should work given an `obj`. As 30 days of continous work on the project happened, I simply accepted that this is not how sizeof works, did ever work, or should work. The line you quoted is simply wrong. I didnt write it.
